File: youtube/apis/playlists.py
"""Playlists API namespace.

Wraps the following YouTube endpoints:

* GET    /playlists — list playlists
* POST   /playlists — create a playlist
* PUT    /playlists — update a playlist
* DELETE /playlists — delete a playlist

Reference: https://developers.google.com/youtube/v3/docs/playlists
"""

from __future__ import annotations

from collections.abc import AsyncIterator
from typing import Any

from youtube.apis.base import BaseAPI
from youtube.config import YOUTUBE_API_BASE_URL
from youtube.models.playlists import (
    Playlist,
    PlaylistListResponse,
    PlaylistPart,
)

_BASE = YOUTUBE_API_BASE_URL


class PlaylistsAPI(BaseAPI):
    """Methods for managing YouTube playlists."""

    async def list(
        self,
        *,
        parts: list[PlaylistPart],
        channel_id: str | None = None,
        id: str | list[str] | None = None,
        mine: bool | None = None,
        max_results: int = 5,
        page_token: str | None = None,
        hl: str | None = None,
    ) -> PlaylistListResponse:
        """Return a list of playlists matching the given filter."""
        params: dict[str, Any] = {
            "part": _join(parts),
            "maxResults": max_results,
        }
        if channel_id is not None:
            params["channelId"] = channel_id
        if id is not None:
            params["id"] = ",".join(id) if isinstance(id, list) else id
        if mine is not None:
            params["mine"] = str(mine).lower()
        if page_token is not None:
            params["pageToken"] = page_token
        if hl is not None:
            params["hl"] = hl

        payload = await self._session.get(f"{_BASE}/playlists", params=params)
        return PlaylistListResponse.model_validate(payload)

    async def iter_mine(
        self,
        *,
        parts: list[PlaylistPart],
        max_results: int = 25,
    ) -> AsyncIterator[Playlist]:
        """Async generator that pages through the authenticated user's playlists."""
        page_token: str | None = None
        while True:
            page = await self.list(parts=parts, mine=True, max_results=max_results, page_token=page_token)
            for item in page.items:
                yield item
            if not page.next_page_token:
                break
            page_token = page.next_page_token

    async def insert(
        self,
        body: Playlist,
        *,
        parts: list[PlaylistPart] | None = None,
    ) -> Playlist:
        """Create a new playlist."""
        if parts is None:
            parts = _infer_parts(body)

        params: dict[str, Any] = {"part": _join(parts)}
        payload = await self._session.post(
            f"{_BASE}/playlists",
            json=body.model_dump(by_alias=True, exclude_none=True),
            params=params,
        )
        return Playlist.model_validate(payload)

    async def update(
        self,
        body: Playlist,
        *,
        parts: list[PlaylistPart] | None = None,
    ) -> Playlist:
        """Update an existing playlist.

        The ``body`` must contain ``id`` and ``snippet.title``.
        """
        if parts is None:
            parts = _infer_parts(body)

        params: dict[str, Any] = {"part": _join(parts)}
        payload = await self._session.put(
            f"{_BASE}/playlists",
            json=body.model_dump(by_alias=True, exclude_none=True),
            params=params,
        )
        return Playlist.model_validate(payload)

    async def delete(self, *, id: str) -> None:
        """Delete a playlist."""
        await self._session.delete(f"{_BASE}/playlists", params={"id": id})


def _join(parts: list[PlaylistPart]) -> str:
    return ",".join(p.value for p in parts)


def _infer_parts(body: Playlist) -> list[PlaylistPart]:
    parts: list[PlaylistPart] = []
    if body.snippet is not None:
        parts.append(PlaylistPart.SNIPPET)
    if body.status is not None:
        parts.append(PlaylistPart.STATUS)
    if body.localizations is not None:
        parts.append(PlaylistPart.LOCALIZATIONS)
    return parts or [PlaylistPart.SNIPPET]

File: go/playlists.go
package youtube

import (
	"context"
	"errors"
	"net/http"
	"net/url"
	"strconv"
	"strings"
)

// Playlist parts accepted by the part parameter.
const (
	PlaylistPartID             = "id"
	PlaylistPartSnippet        = "snippet"
	PlaylistPartStatus         = "status"
	PlaylistPartContentDetails = "contentDetails"
	PlaylistPartPlayer         = "player"
	PlaylistPartLocalizations  = "localizations"
)

// PlaylistSnippet is the snippet part of a Playlist resource.
type PlaylistSnippet struct {
	PublishedAt     string            `json:"publishedAt,omitempty"`
	ChannelID       string            `json:"channelId,omitempty"`
	Title           string            `json:"title,omitempty"`
	Description     string            `json:"description,omitempty"`
	Thumbnails      *ThumbnailDetails `json:"thumbnails,omitempty"`
	ChannelTitle    string            `json:"channelTitle,omitempty"`
	DefaultLanguage string            `json:"defaultLanguage,omitempty"`
	Localized       *Localization     `json:"localized,omitempty"`
}

// PlaylistStatus is the status part of a Playlist resource.
type PlaylistStatus struct {
	PrivacyStatus string `json:"privacyStatus,omitempty"`
}

// PlaylistContentDetails is the contentDetails part of a Playlist resource.
type PlaylistContentDetails struct {
	ItemCount *int `json:"itemCount,omitempty"`
}

// PlaylistPlayer is the player part of a Playlist resource.
type PlaylistPlayer struct {
	EmbedHTML string `json:"embedHtml,omitempty"`
}

// Playlist is a YouTube Playlist resource (request body and response).
type Playlist struct {
	Kind           string                  `json:"kind,omitempty"`
	Etag           string                  `json:"etag,omitempty"`
	ID             string                  `json:"id,omitempty"`
	Snippet        *PlaylistSnippet        `json:"snippet,omitempty"`
	Status         *PlaylistStatus         `json:"status,omitempty"`
	ContentDetails *PlaylistContentDetails `json:"contentDetails,omitempty"`
	Player         *PlaylistPlayer         `json:"player,omitempty"`
	Localizations  map[string]Localization `json:"localizations,omitempty"`
}

// PlaylistListResponse is the response from ListPlaylists.
type PlaylistListResponse struct {
	Kind          string     `json:"kind,omitempty"`
	Etag          string     `json:"etag,omitempty"`
	NextPageToken string     `json:"nextPageToken,omitempty"`
	PrevPageToken string     `json:"prevPageToken,omitempty"`
	PageInfo      *PageInfo  `json:"pageInfo,omitempty"`
	Items         []Playlist `json:"items"`
}

// ListPlaylistsParams configures ListPlaylists.
type ListPlaylistsParams struct {
	// Parts is the list of resource parts to include. Required.
	Parts []string
	// ChannelID returns playlists owned by the channel.
	ChannelID string
	// IDs restricts the result to specific playlist IDs.
	IDs []string
	// Mine, when non-nil, returns the authenticated user's playlists.
	Mine *bool
	// MaxResults is the page size. Defaults to 5.
	MaxResults int
	// PageToken is the pagination cursor.
	PageToken string
	// HL is the language for localized metadata.
	HL string
}

// ListPlaylists returns playlists matching the given filter. Mirrors
// playlists.list.
func (c *Client) ListPlaylists(ctx context.Context, p ListPlaylistsParams) (*PlaylistListResponse, error) {
	if len(p.Parts) == 0 {
		return nil, errors.New("youtube: ListPlaylists requires at least one part")
	}
	maxResults := p.MaxResults
	if maxResults <= 0 {
		maxResults = 5
	}
	q := url.Values{}
	q.Set("part", strings.Join(p.Parts, ","))
	q.Set("maxResults", strconv.Itoa(maxResults))
	if p.ChannelID != "" {
		q.Set("channelId", p.ChannelID)
	}
	if len(p.IDs) > 0 {
		q.Set("id", strings.Join(p.IDs, ","))
	}
	if p.Mine != nil {
		q.Set("mine", strconv.FormatBool(*p.Mine))
	}
	if p.PageToken != "" {
		q.Set("pageToken", p.PageToken)
	}
	if p.HL != "" {
		q.Set("hl", p.HL)
	}

	out := &PlaylistListResponse{}
	if err := c.doGet(ctx, "playlists", q, out); err != nil {
		return nil, err
	}
	return out, nil
}

// IterMyPlaylists walks every playlist owned by the authenticated user,
// invoking fn for each. It pages through the result set until exhausted or fn
// returns an error. This is the Go-idiomatic equivalent of the Python iter_mine
// async generator.
func (c *Client) IterMyPlaylists(
	ctx context.Context,
	parts []string,
	maxResults int,
	fn func(Playlist) error,
) error {
	if maxResults <= 0 {
		maxResults = 25
	}
	mine := true
	pageToken := ""
	for {
		page, err := c.ListPlaylists(ctx, ListPlaylistsParams{
			Parts:      parts,
			Mine:       &mine,
			MaxResults: maxResults,
			PageToken:  pageToken,
		})
		if err != nil {
			return err
		}
		for _, item := range page.Items {
			if err := fn(item); err != nil {
				return err
			}
		}
		if page.NextPageToken == "" {
			return nil
		}
		pageToken = page.NextPageToken
	}
}

// InsertPlaylist creates a new playlist. When parts is empty, the parts are
// inferred from the populated fields of the body. Mirrors playlists.insert.
// Requires OAuth.
func (c *Client) InsertPlaylist(ctx context.Context, body Playlist, parts []string) (*Playlist, error) {
	if len(parts) == 0 {
		parts = inferPlaylistParts(body)
	}
	q := url.Values{}
	q.Set("part", strings.Join(parts, ","))

	out := &Playlist{}
	if err := c.doJSON(ctx, http.MethodPost, "playlists", q, body, out); err != nil {
		return nil, err
	}
	return out, nil
}

// UpdatePlaylist updates an existing playlist. The body must contain id and
// snippet.title. When parts is empty, the parts are inferred from the populated
// fields of the body. Mirrors playlists.update. Requires OAuth.
func (c *Client) UpdatePlaylist(ctx context.Context, body Playlist, parts []string) (*Playlist, error) {
	if len(parts) == 0 {
		parts = inferPlaylistParts(body)
	}
	q := url.Values{}
	q.Set("part", strings.Join(parts, ","))

	out := &Playlist{}
	if err := c.doJSON(ctx, http.MethodPut, "playlists", q, body, out); err != nil {
		return nil, err
	}
	return out, nil
}

// DeletePlaylist deletes a playlist by ID. Mirrors playlists.delete. Requires
// OAuth.
func (c *Client) DeletePlaylist(ctx context.Context, id string) error {
	if id == "" {
		return errors.New("youtube: DeletePlaylist requires an ID")
	}
	q := url.Values{}
	q.Set("id", id)
	return c.doDelete(ctx, "playlists", q)
}

// inferPlaylistParts derives the resource parts from the populated fields of a
// Playlist body, mirroring the Python _infer_parts helper.
func inferPlaylistParts(body Playlist) []string {
	parts := make([]string, 0, 3)
	if body.Snippet != nil {
		parts = append(parts, PlaylistPartSnippet)
	}
	if body.Status != nil {
		parts = append(parts, PlaylistPartStatus)
	}
	if body.Localizations != nil {
		parts = append(parts, PlaylistPartLocalizations)
	}
	if len(parts) == 0 {
		return []string{PlaylistPartSnippet}
	}
	return parts
}
